plugins/tech-insights-backend/src/service/FactRetrieverEngine.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { TechInsightsStore } from './TechInsightsDatabase';
+import { FactRetrieverRegistry } from './FactRetrieverRegistry';
+import cron, { ScheduledTask } from 'node-cron';
+import { FactRetrieverContext } from '../types';
+
+function randomInt(min: number, max: number) {
+  // min and max included
+  return Math.floor(Math.random() * (max - min + 1) + min);
+}
+
+export class FactRetrieverEngine {
+  private scheduledJobs = new Map<string, ScheduledTask>();
+
+  static async create(
+    repository: TechInsightsStore,
+    factRetrieverRegistry: FactRetrieverRegistry,
+    factRetrieverContext: FactRetrieverContext,
+  ) {
+    await Promise.all(
+      factRetrieverRegistry
+        .list()
+        .map(it => repository.insertFactSchema(it.ref, it.schema)),
+    );
+
+    return new FactRetrieverEngine(
+      repository,
+      factRetrieverRegistry,
+      factRetrieverContext,
+    );
+  }
+
+  constructor(
+    readonly repository: TechInsightsStore,
+    readonly factRetrieverRegistry: FactRetrieverRegistry,
+    readonly factRetrieverContext: FactRetrieverContext,
+  ) {}
+
+  schedule() {
+    const retrievers = this.factRetrieverRegistry.list();
+    retrievers.forEach(retriever => {
+      if (!this.scheduledJobs.has(retriever.ref)) {
+        const randomDailyCron = `${randomInt(0, 59)} ${randomInt(0, 23)} * * *`;
+        const job = cron.schedule(
+          retriever.cadence || randomDailyCron,
+          async () => {
+            const facts = await retriever.handler(this.factRetrieverContext);
```

Comment:
Related to my other comment, and perhaps this is too much of an implementation detail (but I feel it might need interface changes), but I wonder if we need to consider processing like the catalog? i.e. base scheduling more around a specific entities than around the retriever. The schedule might mean that we run the retriever once per day for a given entity (or according to it's schedule) perhaps by issuing an event once a day (highly speculative). Our scheduler could also respond to external events and schedule retrievals on demand too.

Sorry, again, if this is too much into implementation.

Reply:
I think the retriever should decide when it needs data. There could be polling cases where the cadence is every minute (for example storing facts about GitHub rate limits) or it could be that it needs to be triggered every month (for example retrieving the results of a month long sprint from Jira). Now that I think about it I do think that the cadence should be defined outside of the retriever though.